docs/proposals/garbage-collection.md

```diff
@@ -0,0 +1,392 @@
+
+## API changes
+
+```
+type DeleteOptions struct {
+	…
+	OrphanDependents bool
+}
+```
+
+**DeleteOptions.OrphanDependents**: allows a user to express whether the dependent objects should be orphaned. It defaults to true, because controllers before release 1.2 expect dependent objects to be orphaned.
+
+## Changes to existing components
+
+* API server:
+When handling a deletion request, depending on if DeleteOptions.OrphanDependents is true, the API server updates the object to add/remove the "orphan" finalizer to/from the ObjectMeta.Finalizers map.
+
+
+## New components
+
+Adding a fourth component to the Garbage Collector, the"orphan" finalizer:
+* Watches for update events as described in [Part I](#part-i-the-finalizer-framework).
+* Removes the object in the event from the `OwnerReferences` of its dependents.
+  * dependent objects can be found via the DAG kept by the GC, or by relisting the dependent resource and checking the OwnerReferences field of each potential dependent object.
+* Also removes any dangling owner references the dependent objects have.
+* At last, removes the itself from the `ObjectMeta.Finalizers` of the object.
+
+# Related issues
+
+## Orphan adoption
+
+Controllers are responsible for adopting orphaned dependent resources. To do so, controllers
+* Checks a potential dependent object’s OwnerReferences to determine if it is orphaned.
+* Fills the OwnerReferences if the object matches the controller’s selector and is orphaned.
+
+There is a potential race between the "orphan" finalizer removing an owner reference and the controllers adding it back during adoption. Imagining this case: a user deletes an owning object and intends to orphan the dependent objects, so the GC removes the owner from the dependent object's OwnerReferences list, but the controller of the owner resource hasn't observed the deletion yet, so it adopts the dependent again and adds the reference back, resulting in the mistaken deletion of the dependent object. This race can be avoided by implementing Status.ObservedGeneration in all resources. Before updating the dependent Object's OwnerReferences, the "orphan" finalizer checks Status.ObservedGeneration of the owning object to ensure its controller has already observed the deletion.
+
+## Interaction with graceful deletion
+
+If the user sets the DeleteOptions.GracePeriodSeconds when deleting an owning object, the garbage collector should apply the same grace period when deleting the dependent objects. To persist the grace period set by the user, the owning object should not be deleted from the registry until all its dependent objects are in the graceful deletion state. This could be ensured by introducing another finalizer, tentatively named as the "populating graceful deletion" finalizer. Upon receiving the graceful deletion request, the API server adds this finalizer to the finalizers list of the owning object. Later the GC will remove it when all dependents are in the graceful deletion state.
```

Comment:
> What if we start this paragraph with something like, "when we want to make graceful deletion implemented for more than just pods and propagate from controllers to the things they control, then..."

Sure.

> Correct me if I'm wrong, but today deleting an RC with grace period 0 doesn't do anything?

You are right.

> Actually, @caesarxuchao, I think Clayton was saying that if kubelet deletes a pod with grace period = 0 and something is owned by the pod and the pod wants to orphan that thing, then kubelet might have a problem. This is a lot different than making grace period propagate. I don't want to make grace period propagate as part of this effort.

I was talking about this comment: https://github.com/kubernetes/kubernetes/pull/23656#issuecomment-215612461. I think Clayton did mean to propagate grace period.